// minecrafthawkeye/dist/types/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.weaponsProps = exports.isEntity = exports.Weapons = exports.Vec2 = void 0;
let Weapons = exports.Weapons = /*#__PURE__*/function (Weapons) {
  Weapons["bow"] = "bow";
  Weapons["crossbow"] = "crossbow";
  Weapons["trident"] = "trident";
  Weapons["ender_pearl"] = "ender_pearl";
  Weapons["snowball"] = "snowball";
  Weapons["egg"] = "egg";
  Weapons["splash_potion"] = "splash_potion";
  return Weapons;
}({});
const weaponsProps = exports.weaponsProps = {
  bow: {
    BaseVo: 3,
    GRAVITY: 0.05,
    waitTime: 1200
  },
  crossbow: {
    BaseVo: 3.15,
    GRAVITY: 0.05,
    waitTime: 1200
  },
  trident: {
    BaseVo: 2.5,
    GRAVITY: 0.05,
    waitTime: 1200
  },
  ender_pearl: {
    BaseVo: 1.5,
    GRAVITY: 0.03,
    waitTime: 150
  },
  snowball: {
    BaseVo: 1.5,
    GRAVITY: 0.03,
    waitTime: 150
  },
  egg: {
    BaseVo: 1.5,
    GRAVITY: 0.03,
    waitTime: 150
  },
  splash_potion: {
    BaseVo: 0.4,
    GRAVITY: 0.03,
    waitTime: 150
  }
};
const isEntity = e => {
  return "type" in e;
};
exports.isEntity = isEntity;
class Vec2 {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }
}
exports.Vec2 = Vec2;
