// minecrafthawkeye/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {};
exports.default = void 0;
var _mineflayer = require("mineflayer");
var _hawkEye = require("./hawkEye");
var _botFunctions = require("./botFunctions");
var _hawkEyeEquations = _interopRequireDefault(require("./hawkEyeEquations"));
var _loadBot = _interopRequireDefault(require("./loadBot"));
var _projectilRadar = require("./projectilRadar");
var _calculateArrowTrayectory = require("./calculateArrowTrayectory");
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const inject = bot => {
  (0, _loadBot.default)(bot);
  bot.hawkEye = {
    oneShot: (from, weapon = _types.Weapons.bow) => (0, _hawkEye.autoAttack)(from, weapon, true),
    autoAttack: (from, weapon = _types.Weapons.bow) => (0, _hawkEye.autoAttack)(from, weapon, false),
    getMasterGrade: _hawkEyeEquations.default,
    simplyShot: _botFunctions.simplyShot,
    stop: _hawkEye.stop,
    getPlayer: _botFunctions.getPlayer,
    calculateArrowTrayectory: _calculateArrowTrayectory.calculateArrowTrayectory,
    detectProjectiles: _hawkEye.detectProjectiles,
    detectAim: _projectilRadar.detectAim,
    startRadar: _projectilRadar.startRadar,
    stopRadar: _projectilRadar.stopRadar
  };
};
var _default = exports.default = inject;
